### PR TITLE
Wip/meta language - class syntax

### PR DESCRIPTION
**Pull request by rayslxu**

### CODE REVIEW

File: lib/class_def_ast.js [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+            metadata, annotations);
+    }
+
+    toString() {
+        let annotations = '';
+        Object.entries(this.metadata).forEach(([name, value]) => {
+            annotations += ` #_[${name}=${prettyprintValue(value)}]`;
+        });
+        Object.entries(this.annotations).forEach(([name, value]) => {
+            annotations += ` #_[${name}=${prettyprintValue(value)}]`;
```
Comment (gcampax) [inline lib/class_def_ast.js]: This should just be `#[]`
(we might need better codegen tests...)

File: lib/class_def_ast.js [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        return this._annotations;
+    }
+
+    toString(prefix = '') {
+        let annotations = '';
+        Object.entries(this.metadata).forEach(([name, value]) => {
+            annotations += `\n${prefix}#_[${name}=${prettyprintValue(value)}]`;
+        });
+        Object.entries(this.annotations).forEach(([name, value]) => {
+            annotations += `\n${prefix}#_[${name}=${prettyprintValue(value)}]`;
```
Comment (gcampax) [inline lib/class_def_ast.js]: Same here

File: lib/grammar.pegjs [hunk 3]
```diff
@@ -503,21 +538,44 @@ array_value_list = first:value _ rest:(',' _ value _)* {
...
+    return { type: 'nl', name: name, value: value };
+}
+impl_annotation = '#' '[' name:ident _ '=' value:value ']' {
+    return { type: 'impl', name: name, value: value };
+}
+
 // Types
 
-type_ref = 'Measure' _ '(' _ unit:ident? _ ')' { return Type.Measure(unit); } /
+type_ref = hashable_type_ref /
```
Comment (gcampax) [inline lib/grammar.pegjs]: do you still use this distinction?

File: lib/utils.js [hunk 4]
```diff
@@ -86,6 +86,10 @@ module.exports = {
             return schemaRetriever.getSchemaAndNames(type, schemaType, name).then((schema) => module.exports.splitArgsForSchema(schema, schemaType, false));
     },
 
+    getMixin(schemaRetriever, kind) {
+        return schemaRetriever.getMixins(kind);
+    },
+
```
Comment (gcampax) [inline lib/utils.js]: do you need this indirection?

File: lib/grammar.pegjs (hunk 3 above)
Comment (gcampax) [reply lib/grammar.pegjs]: if we don't need it, it's just extra cognitive load on our part when we read this code later on; as in, we'll come back 3 months from now to this code and wonder: "what was this again?" "is my new type hashable or not?"

File: lib/utils.js (hunk 4 above)
Comment (rayslxu) [reply lib/utils.js]: just trying to be consistent with getschema, is it bad?

File: lib/class_def_ast.js (hunk 1 above)
Comment (rayslxu) [reply lib/class_def_ast.js]: fixed

File: lib/class_def_ast.js (hunk 2 above)
Comment (rayslxu) [reply lib/class_def_ast.js]: fixed

File: lib/utils.js (hunk 4 above)
Comment (gcampax) [reply lib/utils.js]: It's extra code that does not add much to readability. So yeah I'd get rid of it.

**Pull request merged**